utilize SOURCE_VERSION variable for fingerprinting

jackdwyer opened the pull request:
Write `SOURCE_VERSION` to `.profile.d` so it can be sourced on boot.

Not sure if this is best:
```
slug_fingerprint = ENV['SOURCE_VERSION'] || abort('no SOURCE_VERSION variable avaliable')
```
should I set the value to eg: `NONE` 

@HoyaBoya @blueapron/devops 

```bash
$ heroku run console -a ba-wms-staging-5
Running console on ba-wms-staging-5... up, run.6922
Loading staging environment (Rails 4.2.5.1)
[1] pry(main)> ENV['SOURCE_VERSION']
=> "5e01ada562678267019e5a99b5b2b5d817f02442"
```

```bash
$ heroku run bash -a ba-wms-staging-5
Running bash on ba-wms-staging-5... up, run.8403
~ $ cat .profile.d/source_version.sh 
export SOURCE_VERSION=5e01ada562678267019e5a99b5b2b5d817f02442
~ $ env | grep SOURCE_VERSION
SOURCE_VERSION=5e01ada562678267019e5a99b5b2b5d817f02442
```

HoyaBoya commented on bin/compile:
```diff
@@ -16,3 +16,9 @@ LanguagePack::Instrument.trace 'compile', 'app.compile' do
   end
 end
+
+slug_fingerprint = ENV['SOURCE_VERSION'] || abort('no SOURCE_VERSION variable avaliable')
```
Shouldn't the `ENV` automatically fail here and not get to the `abort`. I would think `fetch('SOURCE_VERSION', nil)` would get to your more gracious `abort`.

HoyaBoya commented:
Just to clarify my next steps if I understand this PR, I'll then be able to serve this up via any `ENV['SOURCE_VERSION'] command?

jackdwyer commented:
I used `abort` as it should force an exit status of 1, which will fail the
build.

To clarify do we want the build to fail if we can not use SOURCE_VERSION?
I might update to call out to `git rev-list HEAD` if the env is not set.

Correct the environment variable should be avaliable with
`ENV['SOURCE_VERSION']` if built with this buildpack
On 29 Feb 2016 12:53 pm, "HoyaBoya" <notifications@github.com> wrote:

> Just to clarify my next steps if I understand this PR, I'll then be able
> to serve this up via any `ENV['SOURCE_VERSION'] command?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/blueapron/heroku-buildpack-ruby/pull/2#issuecomment-190309285>
> .
>

HoyaBoya commented:
I am corrected, I forgot my own Ruby.

HoyaBoya commented:
I think this makes sense to me if it loads the hash into `ENV`. I am :+1: but defer also to @prsimp as well.

prsimp commented:
Just for reference, the `SOURCE_VERSION` variable is added during the build process by Heroku ([link](https://devcenter.heroku.com/changelog-items/630)):

> A `SOURCE_VERSION` environment variable is now available to builds running on Heroku. For git-push builds, this is the git commit SHA-1 of the source being built. Examples of `SOURCE_VERSION` uses are versioning generated assets or fingerprinting slugs.

Looks good to me, :+1:

jackdwyer commented:
added a call to git just in case.

HoyaBoya commented:
> added a call to git just in case.

Could this "backup" call lead to a split brain problem? Might be better for all apps to do it one way or explode.

HoyaBoya commented:
PR is here for the middleware that will use this work.

https://github.com/blueapron/blueapron.com/pull/4565

jackdwyer commented:
@HoyaBoya we do not actually have access to `git` inside the buildpack.  As such I have removed the shell out.
I have reverted to using `abort`.
Anything built outside of Heroku, I will  expose the `SOURCE_VERSION` environment variable. It will be generated by `git rev-parse --verify HEAD`

HoyaBoya commented:
Awesome @jackdwyer. I think the :+1: from @prsimp still stands with the revert. Take a look at the middleware PR real quick, and just let me know if that's what I should be reading out.

Nice stuff dude.

HoyaBoya commented:
> Anything built outside of Heroku, I will expose the SOURCE_VERSION environment variable. It will be generated by git rev-parse --verify HEAD

Makes sense, hence why people have all these hooks. Just as long as Heroku and EC2 are pushing into `ENV['SOURCE_VERSION']` we should be solid with the Middleware.

:+1:
